### PR TITLE
Smaller fixes to the backprop operation & lowering

### PR DESCRIPTION
**Pull request by dime10**
Some improvements & fixes:
 - bug fix in the frontend to allow `grad` on `qjit(f)` where `f` is not a QNode
 - some refactoring of the backprop op
 - new verifiers for the backprop op
 - switch to allocating primal output buffers during bufferization rather than backprop -> llvm lowering

### CODE REVIEW

File: mlir/lib/Gradient/Transforms/BufferizationPatterns.cpp [hunk 1]
```diff
@@ -15,18 +15,46 @@
...
 
+/// Helper function to generate a set of memref allocations.
+///
+/// The allocation size and shape is deduced from a list of existing memref values.
+///
+void generateAllocations(PatternRewriter &rewriter, Location loc,
+                         SmallVectorImpl<Value> &allocations, ValueRange referenceValues)
+{
+    for (Value memref : referenceValues) {
+        MemRefType memrefType = memref.getType().cast<MemRefType>();
```
Comment (pengmai) [inline mlir/lib/Gradient/Transforms/BufferizationPatterns.cpp:40]: nit: this style of casting is going to be removed
```suggestion
        MemRefType memrefType = cast<MemRefType>(memref.getType());
```

**Review by pengmai — APPROVED** [review]
Looks great!

File: mlir/lib/Gradient/Transforms/BufferizationPatterns.cpp (hunk 1 above)
Comment (dime10) [reply mlir/lib/Gradient/Transforms/BufferizationPatterns.cpp:40]: Good point, thanks!

File: frontend/test/pytest/test_gradient.py [hunk 2]
```diff
@@ -108,6 +108,19 @@ def workflow(p: float):
...
 
 
+def test_grad_on_qjit():
+    """Check that grad works when called on an existing qjit object that does not wrap a QNode."""
+
+    @qjit
+    def f(x: float):
+        return x * x
+
+    result = qjit(grad(f))(3.0)
```
Comment (rmoyard) [inline frontend/test/pytest/test_gradient.py:118]: Nice :+1:

File: mlir/include/Gradient/IR/GradientOps.td [hunk 3]
```diff
@@ -88,28 +88,33 @@ def AdjointOp : Gradient_Op<"adjoint", [AttrSizedOperandSegments]> {
...
+        DeclareOpInterfaceMethods<SymbolUserOpInterface>]> {
     let summary = "Perform classic automatic differentiation using Enzyme AD.";
 
     let arguments = (ins
         FlatSymbolRefAttr:$callee,
         Variadic<AnyType>:$args,
-        Variadic<AnyType>:$quantum_jacobian,
-        Variadic<AnyMemRef>:$data_in,
+        Variadic<MemRefOf<[AnyFloat]>>:$diffArgShadows,
+        Variadic<MemRefOf<[AnyFloat]>>:$calleeResults,
```
Comment (rmoyard) [inline mlir/include/Gradient/IR/GradientOps.td:99]: Why did we need the callee results here?

File: mlir/lib/Gradient/Transforms/BufferizationPatterns.cpp [hunk 4]
```diff
@@ -61,24 +89,28 @@ class BufferizeBackpropOp : public OpConversionPattern<BackpropOp> {
...
+        const std::vector<Value> &diffArgs =
+            computeDiffArgs(adaptor.getArgs(), op.getDiffArgIndices());
+        generateAllocations(rewriter, loc, argShadows, diffArgs);
+
+        // Enzyme requires buffers for the primal outputs as well, even though we don't need their
+        // values. We'll mark them dupNoNeed later on to allow Enzyme to optimize away their
+        // computation.
+        SmallVector<Value> calleeResults;
+        ValueRange resShadows = adaptor.getCotangents();
+        generateAllocations(rewriter, loc, calleeResults, resShadows);
```
Comment (rmoyard) [inline mlir/lib/Gradient/Transforms/BufferizationPatterns.cpp:107]: :+1: A lot cleaner

**Review by rmoyard — APPROVED** [review]
Thanks great PR :+1:

File: mlir/include/Gradient/IR/GradientOps.td (hunk 3 above)
Comment (dime10) [reply mlir/include/Gradient/IR/GradientOps.td:99]: The idea was to handle all allocations during the bufferization stage. At the moment the buffers for the callee results (required by Enzyme) were allocated during the lowering, but I think it makes more sense during the bufferization stage.

Comment (pengmai) [reply mlir/include/Gradient/IR/GradientOps.td:99]: Enzyme requires allocated space for the primal (callee) results even though we don't need their values. Adding them as arguments here allows us to allocate this space during bufferization, which feels more appropriate than allocating them during conversion to the LLVM dialect.

Comment (rmoyard) [reply mlir/include/Gradient/IR/GradientOps.td:99]: Thanks a lot :+1:

**Pull request merged**